pkg/server/sandbox_stop.go

```diff
@@ -77,16 +78,58 @@ func (c *criContainerdService) StopPodSandbox(ctx context.Context, r *runtime.St
 	glog.V(2).Infof("TearDown network for sandbox %q successfully", id)
 
 	sandboxRoot := getSandboxRootDir(c.rootDir, id)
-	if err = c.unmountSandboxFiles(sandboxRoot, sandbox.Config); err != nil {
+	if err := c.unmountSandboxFiles(sandboxRoot, sandbox.Config); err != nil {
 		return nil, fmt.Errorf("failed to unmount sandbox files in %q: %v", sandboxRoot, err)
 	}
 
-	// TODO(random-liu): [P1] Handle sandbox container graceful deletion.
+	if err := c.stopSandboxContainer(ctx, id); err != nil {
+		return nil, fmt.Errorf("failed to stop sandbox container %q: %v", id, err)
+	}
+	return &runtime.StopPodSandboxResponse{}, nil
+}
+
+// stopSandboxContainer kills and deletes sandbox container.
+func (c *criContainerdService) stopSandboxContainer(ctx context.Context, id string) error {
+	if _, err := c.taskService.Kill(ctx, &tasks.KillRequest{
+		ContainerID: id,
+		Signal:      uint32(unix.SIGKILL),
+		All:         true,
+	}); err != nil && !isContainerdGRPCNotFoundError(err) && !isRuncProcessAlreadyFinishedError(err) {
+		return fmt.Errorf("failed to kill sandbox container: %v", err)
+	}
+
+	if err := c.waitSandboxContainer(ctx, id); err != nil && !isContainerdGRPCNotFoundError(err) {
+		return fmt.Errorf("failed to wait for pod sandbox: %v", err)
+	}
+
 	// Delete the sandbox container from containerd.
-	_, err = c.taskService.Delete(ctx, &execution.DeleteRequest{ContainerID: id})
+	_, err := c.taskService.Delete(ctx, &tasks.DeleteTaskRequest{ContainerID: id})
 	if err != nil && !isContainerdGRPCNotFoundError(err) {
-		return nil, fmt.Errorf("failed to delete sandbox container %q: %v", id, err)
+		return fmt.Errorf("failed to delete sandbox container: %v", err)
 	}
 
-	return &runtime.StopPodSandboxResponse{}, nil
+	return nil
+}
+
+// waitSandboxContainer polls sandbox container status until it's stopped.
+// TODO(random-liu): Rely on event after switching to containerd client.
+func (c *criContainerdService) waitSandboxContainer(ctx context.Context, id string) error {
```

Comment:
Using event is a bit complex here. Since we are switching to containerd client soon. Just use polling for now.